README
Bunny, Mayoress, Prostitute & Succubus (BMPS) - Game

Type: Text Adventure/Interactive Fiction
Version: 0.82 (Wednesday 1st February 2017)
Platforms: Linux and Windows command prompt mini-web-server, accessed with a web browser.
Original Author: M100 (M100-A)
Images by: Inkeranon (http://notreallybuttpants.tumblr.com/)
Synopsis: The mayor of Belchwood hired a new prostitute for his village, so he could watch his fellow male villagers reduce yet another unfortunate woman to being a mind broken sex slave. He accidentally hired a succubus for the job. She turned the mayor to being her obedient servant. Bunny then took over from the mayor, and now is now plotting to break the other 63 men of Belchwood.

Github note: There is a file called BMPSfiles.dat, which contains all of the images that the mini-web-server serves. However, given that the file contains pornographic images, I should not be uploading that to github.


# Usage (run) instructions:

On Linux:
First you need to compile the program (see below).
Once the program is compiled, run the program from a command line (xterm or SSH session).

Note: There is no Linux binary distributed, so you will need to compile the source code.
Note2: I have found "differences" between various Linux distributions in relation to linkage to shared library (.so) files that cause programs to not necessarily be as cross compatible. I am not taking my chances, so I am just supplying the source code.


On Windows:
Run the program from either a command line or click on the program.
Note: a Windows binary is available.

When the program starts, it will output something like this:
 BMPS Version 0.80 is starting.
 BMPSfiles.dat Catalog = 331 bytes
 BMPSfiles.dat CatalogEntries = 7
 BMPSfiles.dat checks out ok.
 Schedule: 64 NPCs loaded and checked.
 bound & listening.
 Connect with a web browser to this URL:
 http://127.0.0.1:6081/
 Use Ctrl-C to stop the program.

(As of 0.81, on Windows, the program will attempt to start a web browser for you).
The URL supplied is the one that you use a web browser to connect to.


When you have finished with the program, you can press Ctrl-C in the window to close the program.


# System requirements:

On Linux:
gcc compiler
make

Note: you may need to load packages for your Linux distribution for any of the above.


On Windows: (if you want to compile the code)
a C compiler (I used CodeBlocks with MinGW)

Note: I supply the binary for Windows, which should work.


# Compile instructions:

On Linux:
Unpack the files in a directory, navigate to that directory.
Use the "make" command to create the program.
"BMPS" will be created in the directory.


On Windows:
Create a project within you compiler.
Then import the BMPS source code.
You will need to include the WIN32 compile define before compiling.
You will need to link to the ws2_32 library before compiling/linking.
Then start building the program.


# Usage (play) instructions:

When you go to the URL with a web browser, you begin in the main menu, the program has a series of buttons.

The standalone web server version borrows heavily from the CGI-BIN version, so autosave is turned on at all times. The menu mostly matches the CGI-BIN version.

You can also change the "skin" used by the program which dictates the type of font and if the text is black on white, or white on black.


# Command line options:

BMPS (standalone web server version) has 1 optional command line switch (2 if it is windows), and 2 optional parameters.

(Windows only) -W stops the program from starting a web browser upon startup.

-D turns on debug output. In BMPS this just changes the "Give me a hint" primary navigation menu, so that the locations of unknown and partially known NPCs are shown, "progress values" are shown next to the names of the NPCs, and #, * and $ identifiers are used for if the NPC is exhausted, sound asleep or is locked. Also, "Give me a hint" is always available (unlike in non-debug mode, it is unavailable for a while depending upon the difficulty level).

The first optional parameter is the port. The program defaults to 6081, but you can change it at the command line.
For example: ./BMPS 8080

The second optional parameter is the bind IPaddress. The program defaults to 127.0.0.1. However, you need to also specify the port before it.
For example: ./BMPS 8080 127.0.0.2


# Program output:

As you play BMPS via the web browser, the program in its text window will be outputing the log information into the window as you do things.

 Use Ctrl-C to stop the program.
 client 127.0.0.1 18121: GET /, S 200 done.
 client 127.0.0.1 18633: GET /favicon.ico, File ico done.
 client 127.0.0.1 20169: POST /, 17 posted bytes, S 200 done.
 client 127.0.0.1 20681: GET /images/Concept1.jpg, File jpg done.
 client 127.0.0.1 21193: POST /, 19 posted bytes, S 200 done.
 client 127.0.0.1 23241: POST /, 17 posted bytes, S 200 done.
 client 127.0.0.1 23457: POST /, 43 posted bytes, S 200 done.
 client 127.0.0.1 23969: GET /images/StartScene1.jpg, File jpg done.
 client 127.0.0.1 23481: GET /images/StartScene2.jpg, File jpg done.
 client 127.0.0.1 23993: POST /, 43 posted bytes, S 200 done.
 client 127.0.0.1 24753: Invalid HTTP structure.. 500 done.
 client 127.0.0.1 25529: POST /, 43 posted bytes, S 200 done.

If you know your HTTP status codes, you would probably see that 500 an be freaking out (second last line). My only advice is don't.

This sometimes happens when web browsers decide to connect to web servers, then realize after they have connected that they have no immediate use for the connection (this happens when caching is involved). The 500 response is triggered because the mini-web server is waiting for input from the web browser, however the web browser decides that it no longer needs the TCP/IP connection and closes it down. The mini-web server ends the communication by trying to send a 500 response back to the web browser, which explains why the 500 response is logged. The web browser should never see it.


M100



CHANGELOG


# Version 0.82 - Wednesday 1st February 2016

4 more images from Inkeranon.


# Version 0.81 - Sunday 11th December 2016

2 more images from Inkeranon.

And a slight change for the WIN32 standalone mini-web-server. After a discussion on /dgg/ I decided to alter the WIN32 mini-web-server to try to automatically start the default web browser.


# Version 0.80 - Friday 2nd December 2016

This is another significant change to the game. For the "local version", has changed from the CURSES version to a standalone mini-web server. The reason for this is to allow for the images to be shown with the text. A CURSES command line version cannot show images (it is strictly for text).

Version 0.70 save files for the CURSES version are not compatible with Version 0.80 mini-web server version.

The CGI-BIN version has also been updated to have the images.

The images are by Inkeranon (http://notreallybuttpants.tumblr.com).

Spoiler pages have been added to the standalone mini-web server, and CGI-BIN versions.

During testing, I found a bug with IE11, and had to change the way that the javascript works so that clicking on the link next to the radio button actually activates the radio button (Firefox was working correctly).


# Version 0.70 - Friday 11th November 2016

Some grammar and spelling errors have been fixed, but most of the changes were with the underlying code.

As part of the CGI-BIN conversion of this project (BMPS), and my two earlier projects (HQWA and XLEA), I found some faults in some of the lower level functions and those have been corrected.

I have added the CGI-BIN conversion directory to this release.

Note: I did not make the CGI-BIN version cross platform capable.


# Version 0.60 - Wednesday 19th October 2016

Initial alpha version released.

There are likely to be some grammatical errors in the dialog, despite the fact that I have performed a number of QC passes (it is difficult when the developer, writer and editor are the one and the same person).


# Project start - Saturday 24th September 2016.




mdialog.c
//
// mdialog.c
//
// some dialog utilities for the program.
//

#include "common.h"

//
// I put this listing aside from mutils.c, so that I am not altering mutils.
//
// A feature that I am adding into BMPS, is search/replace operations within the dialog.
// I did create a mutils.c function to support this (mutils_addToDialogBufferLenSpec).
//
// This ENTIRE listing is about managing search/replace actions.
//

//
// This is the static search/replace table.
//
// Offset 0 is the PC name
// Offsets 1-64 are the NPC names
// Offset 65 is the Village name
// Offset 66 is a default value (for the dynamic table)
//
struct  mdialog_searchreplacelist gstructSrchRepList[] =
{
    { "00", "Bunny" },
    { "01", "Zeke" },
    { "02", "Otto" },
    { "03", "Lars" },
    { "04", "Nasir" },
    { "05", "Gino" },
    { "06", "Terell" },
    { "07", "Cortez" },
    { "08", "Redinald" },
    { "09", "Sigga" },
    { "10", "Massimo" },
    { "11", "Salvator" },
    { "12", "Kohl" },
    { "13", "Hung" },
    { "14", "Philippe" },
    { "15", "Quon" },
    { "16", "Norris" },
    { "17", "Jamar" },
    { "18", "Tuan" },
    { "19", "Kadeem" },
    { "20", "Erick" },
    { "21", "Izaiah" },
    { "22", "Mack" },
    { "23", "Rishi" },
    { "24", "Jered" },
    { "25", "Bryan" },
    { "26", "Harvir" },
    { "27", "Viktor" },
    { "28", "Cecil" },
    { "29", "Anders" },
    { "30", "Franco" },
    { "31", "Haydn" },
    { "32", "Tarik" },
    { "33", "Arjan" },
    { "34", "Kang" },
    { "35", "Edwin" },
    { "36", "Nate" },
    { "37", "Dalton" },
    { "38", "Rylan" },
    { "39", "Joban" },
    { "40", "Dryden" },
    { "41", "Keyon" },
    { "42", "Lester" },
    { "43", "Yousef" },
    { "44", "Braden" },
    { "45", "Deion" },
    { "46", "Omari" },
    { "47", "Melvin" },
    { "48", "Gary" },
    { "49", "Jerry" },
    { "50", "Frank" },
    { "51", "Pete" },
    { "52", "Lloyd" },
    { "53", "Bob" },
    { "54", "Jake" },
    { "55", "Olson" },
    { "56", "Perry" },
    { "57", "Marcel" },
    { "58", "Rafael" },
    { "59", "Judd" },
    { "60", "Fabian" },
    { "61", "Aho" },
    { "62", "Benson" },
    { "63", "Shaw" },
    { "64", "Javier" },
    { "65", "Belchwood" },
    { "66", "<none>" },
    { NULL, NULL }
};

struct  mdialog_dynamicreplacelist gpstructDynamicReplaceList[MDIALOG_DYNAMIC_MAX + 1];
short   gsSrchRepListMax = 0;

//
// Small function to init the dynamic list.
//
void mdialog_initDynamicList ()
{
    short sOffset1;

    sOffset1 = 0;

    // find the end of the static list

    while (1)
    {
        if (gstructSrchRepList[sOffset1].pszReplace == NULL)
            break;

        sOffset1++;
    }

    gsSrchRepListMax = sOffset1;

    // reset the entire dynamic list
    for (sOffset1 = 0; sOffset1 < MDIALOG_DYNAMIC_MAX; sOffset1++)
    {
        gpstructDynamicReplaceList[sOffset1].sLink = (short)(gsSrchRepListMax - 1);
        sprintf (gpstructDynamicReplaceList[sOffset1].szSearch, "%02d", sOffset1);
    }
}

//
// Small function to setup the dynamic list entry.
//
void mdialog_setDynamicEntry (short sOffset, short sEntry)
{
    if (sOffset >= MDIALOG_DYNAMIC_MAX)
        return;

    if (sEntry >= gsSrchRepListMax)
        return;

    gpstructDynamicReplaceList[sOffset].sLink = sEntry;
}

//
// This is a front end to mutils_addToDialogBuffer & mutils_addToDialogBufferLenSpec,
// which takes care of looking for Search tags and inputting replacements.
//
// One of the grand assumptions is that the input string is null terminated.
//
// 0 if ok
// 1 if there was a memory allocation error
//
int mdialog_addToDialogWithSrchRep (char *pszString)
{
    int iRet;
    int iStart = 0;
    int iPos = 0;
    int iLen = (int)strlen (pszString);
    int iCalc;

    while (1)
    {
        if (iPos >= iLen)
            break;

        if (pszString[iPos] == '#')
        {
            iCalc = iLen - iPos;

            // check that we have enough bytes to lock onto (so we are not potentially searching off the end of the string somewhere).
            if (iCalc > 4)
            {
                if (pszString[iPos + 4] == '#')
                {
                    short sOffset1;
                    char cLock = 0;

                    iCalc = iPos - iStart;

                    //
                    // This is the static search/replace trigger.
                    //
                    if (pszString[iPos + 1] == '!')
                    {
                        for (sOffset1 = 0; sOffset1 < gsSrchRepListMax; sOffset1++)
                        {
                            if (gstructSrchRepList[sOffset1].pszReplace == NULL)
                                break;

                            if (strncmp (gstructSrchRepList[sOffset1].pszSearch, &pszString[iPos + 2], 2) == 0)
                            {
                                cLock = 1;
                                break;
                            }
                        }

                        if (cLock != 0)
                        {
                            // first dump out the string that we have so far
                            iRet = mutils_addToDialogBufferLenSpec (&pszString[iStart], iCalc);
                            if (iRet != 0) return iRet;

                            // now set our position past the search/replace key.
                            iPos += 5;
                            iStart = iPos;

                            // now put in the search/replace value
                            iRet = mutils_addToDialogBuffer (gstructSrchRepList[sOffset1].pszReplace);
                            if (iRet != 0) return iRet;

                            continue;
                        }
                    }

                    //
                    // This is the dynamic search/replace trigger (using what mdialog_setDynamicEntry maintains).
                    //
                    if (pszString[iPos + 1] == '^')
                    {
                        for (sOffset1 = 0; sOffset1 < MDIALOG_DYNAMIC_MAX; sOffset1++)
                        {
                            if (strncmp (gpstructDynamicReplaceList[sOffset1].szSearch, &pszString[iPos + 2], 2) == 0)
                            {
                                cLock = 1;
                                break;
                            }
                        }

                        if (cLock != 0)
                        {
                            // first dump out the string that we have so far
                            iRet = mutils_addToDialogBufferLenSpec (&pszString[iStart], iCalc);
                            if (iRet != 0) return iRet;

                            // now set our position past the search/replace key.
                            iPos += 5;
                            iStart = iPos;

                            // now put in the search/replace value
                            iRet = mutils_addToDialogBuffer (gstructSrchRepList[gpstructDynamicReplaceList[sOffset1].sLink].pszReplace);
                            if (iRet != 0) return iRet;

                            continue;
                        }
                    }
                }
            }
        }
#ifndef INTERFACE_CURSES
        else if (pszString[iPos] == '&')
        {
            iCalc = iPos - iStart;

            // first dump out the string that we have so far
            iRet = mutils_addToDialogBufferLenSpec (&pszString[iStart], iCalc);
            if (iRet != 0) return iRet;

            // now set our new start position.
            iPos++;
            iStart = iPos;

            // insert the replacement character entity.
            iRet = mutils_addToDialogBuffer ("&amp;");
            if (iRet != 0) return iRet;

            continue;
        }
        else if (pszString[iPos] == '<')
        {
            iCalc = iPos - iStart;

            // first dump out the string that we have so far
            iRet = mutils_addToDialogBufferLenSpec (&pszString[iStart], iCalc);
            if (iRet != 0) return iRet;

            // now set our new start position.
            iPos++;
            iStart = iPos;

            // insert the replacement character entity.
            iRet = mutils_addToDialogBuffer ("&lt;");
            if (iRet != 0) return iRet;

            continue;
        }
        else if (pszString[iPos] == '>')
        {
            iCalc = iPos - iStart;

            // first dump out the string that we have so far
            iRet = mutils_addToDialogBufferLenSpec (&pszString[iStart], iCalc);
            if (iRet != 0) return iRet;

            // now set our new start position.
            iPos++;
            iStart = iPos;

            // insert the replacement character entity.
            iRet = mutils_addToDialogBuffer ("&gt;");
            if (iRet != 0) return iRet;

            continue;
        }
        else if (pszString[iPos] == '\"')
        {
            iCalc = iPos - iStart;

            // first dump out the string that we have so far
            iRet = mutils_addToDialogBufferLenSpec (&pszString[iStart], iCalc);
            if (iRet != 0) return iRet;

            // now set our new start position.
            iPos++;
            iStart = iPos;

            // insert the replacement character entity.
            iRet = mutils_addToDialogBuffer ("&quot;");
            if (iRet != 0) return iRet;

            continue;
        }
        else if (pszString[iPos] == '\'')
        {
            iCalc = iPos - iStart;

            // first dump out the string that we have so far
            iRet = mutils_addToDialogBufferLenSpec (&pszString[iStart], iCalc);
            if (iRet != 0) return iRet;

            // now set our new start position.
            iPos++;
            iStart = iPos;

            // insert the replacement character entity.
            iRet = mutils_addToDialogBuffer ("&apos;");
            if (iRet != 0) return iRet;

            continue;
        }
        else if (pszString[iPos] == '\n')
        {
            iCalc = iPos - iStart;

            // first dump out the string that we have so far
            iRet = mutils_addToDialogBufferLenSpec (&pszString[iStart], iCalc);
            if (iRet != 0) return iRet;

            // now set our new start position.
            iStart = iPos;

            // now put in the extra HTML formatting
            iRet = mutils_addToDialogBuffer ("<br/>");
            if (iRet != 0) return iRet;

        }
#endif // INTERFACE_CURSES


        iPos++;
    }

    // finish off the string
    return (mutils_addToDialogBuffer (&pszString[iStart]));
}
//
// Why the logic complexity around search/replace?
//
// The "not much of an issue" that I wanted to avoid was allocating more memory. This is the primary
// reason for the added complexity.
//
// So the idea is to look for the start of the search combinations #! and #^
//
// #! prefixes static S/R tags (#!00#, etc).
// #^ prefixes the dynamic S/R tags (#^00#, etc).
//
// For example:
// 00000000001111111111222222222233333333334
// 01234567890123456789012345678901234567890
// This is a #!00# test of a string.
//
// When I identify that I have a valid S/R combo (complete with a match).
// Let's say for the sake of the demonstration (with the above example), that the code identifies
// this when it reaches offset 10. It checks to see if the remaining string is long enough (iCalc
// check) to be a S/R tag. It then verifies that the last character of the combo is there (trailing
// #, offset 14), then goes looking for if it is a static or dynamic type (because that changes the
// memory table we are looking in).
//
// Then I look in the memory tables by the 2 character string combinations (being the 2 numbers).
// Once a lock has been successfully made:
//  Dump out what was before the SR tag (mutils_addToDialogBufferLenSpec call).
//  Set the current and start positions past the end of the SR tag (which will become offset 15).
//  Dump out the referenced entry in the table lookup (mutils_addToDialogBuffer call).
//  Then resume checking.
//
// At the end of the function is a mutils_addToDialogBuffer call to dump out what is left from the
// updated start position.
//
// The above example is written out as (between the [] brackets).
// [This is a ] mutils_addToDialogBufferLenSpec call
// [Bunny] mutils_addToDialogBuffer call
// [ test of a string.] mutils_addToDialogBuffer call
//
// Any memory allocation is left to the lower level mutils_addToDialogBufferLenSpec and
// mutils_addToDialogBuffer function calls.
//
// It might seem a pain that only character names, and the town name are in the S/R tag combinations.
// However, in this case, the SR values are all defined here and don't change. As a result, if there
// is a need to include the PC, NPC or Town name anywhere, everything can be run through here.
//
// Meaning, if you don't like the names that I came up with for the PC, 64 NPCs or the town, just
// change them here, and the code will use those... as long as I stick to using the S/R tags in the
// first place.
//


